pkg/querier/engine/streaming/operator/aggregation.go

```diff
@@ -0,0 +1,203 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package operator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/util/zeropool"
+)
+
+type Aggregation struct {
+	Inner    InstantVectorOperator
+	Start    time.Time
+	End      time.Time
+	Interval time.Duration
+	Grouping []string
+
+	remainingInnerSeriesToGroup []*group // One entry per series produced by Inner, value is the group for that series
+	remainingGroups             []*group // One entry per group, in the order we want to return them
+}
+
+type group struct {
+	labels labels.Labels
+
+	// The number of input series that belong to this group that we haven't yet seen.
+	remainingSeriesCount uint
+
+	// Sum and presence for each step.
+	sums    []float64
+	present []bool
```

Comment:
Idea for the future: `present []bool` could by replaced by a custom struct which implements it as a bitmap. For example a `[]uint64` where each value is stored in a single bit, so a single `uint64` stores 64 "present" values.

Reply:
Good idea, I've added this to my list of potential performance improvements